Deprecates AdamW and adds `--optim`

manuelciosici opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes #14539


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

@stas00 

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

## Misc
@stas00 
 * I added `FusedAdam` based on your comment in #14539
 * Since both `--optim adafactor` and `--adafactor` server the same purpose, I marked `--adafactor` as deprecated. I copy-pasted a deprecation warning that mentions `transformers` version 5. Let me know if the deprecation warning should say something else.
 * Let me know if I missed anything else.

manuelciosici committed:
Add AdamW deprecation warning

manuelciosici committed:
Add --optim to Trainer

sgugger commented on src/transformers/optimization.py:
```diff
@@ -287,6 +288,8 @@ class AdamW(Optimizer):
             Decoupled weight decay to apply.
         correct_bias (:obj:`bool`, `optional`, defaults to `True`):
             Whether or not to correct bias in Adam (for instance, in Bert TF repository they use :obj:`False`).
+        no_deprecation_warning (:obj:`bool`, `optional`, defaults to `False`):
```
```suggestion
        no_deprecation_warning (:obj:`bool`, `optional`, defaults to :obj:`False`):
```

sgugger commented on src/transformers/optimization.py:
```diff
@@ -287,6 +288,8 @@ class AdamW(Optimizer):
         correct_bias (:obj:`bool`, `optional`, defaults to `True`):
             Whether or not to correct bias in Adam (for instance, in Bert TF repository they use :obj:`False`).
+        no_deprecation_warning (:obj:`bool`, `optional`, defaults to `False`):
+            A flag used to disable the deprecation warning (`True` to disable the warning).
```
```suggestion
            A flag used to disable the deprecation warning (set to :obj:`True` to disable the warning).
```

sgugger commented on src/transformers/optimization.py:
```diff
@@ -297,7 +300,14 @@ def __init__(
         correct_bias: bool = True,
+        no_deprecation_warning: bool = False,
     ):
+        if no_deprecation_warning is False:
```
```suggestion
        if not no_deprecation_warning:
```

sgugger commented on src/transformers/optimization.py:
```diff
@@ -297,7 +300,14 @@ def __init__(
     ):
+        if no_deprecation_warning is False:
+            warnings.warn(
+                "HuggingFace AdamW is deprecated and will be removed in a future version. Use the"
```
```suggestion
                "This implementation of AdamW is deprecated and will be removed in a future version. Use the"
```

sgugger commented on src/transformers/trainer.py:
```diff
@@ -818,17 +818,43 @@ def create_optimizer(self):
+            elif self.args.optim == "adamw_torch":
+                from torch.optim import AdamW
+
+                optimizer_kwargs.update(adam_kwargs)
```
The `optimizer_cls` line is missing in this part of the test.

manuelciosici replied:
Good catch. It should be fixed now.

sgugger commented on src/transformers/training_args.py:
```diff
@@ -330,9 +330,11 @@ class TrainingArguments:
         adafactor (:obj:`bool`, `optional`, defaults to :obj:`False`):
-            Whether or not to use the :class:`~transformers.Adafactor` optimizer instead of
-            :class:`~transformers.AdamW`.
+            This argument is deprecated. Use ``optim`` instead. Whether or not to use the
```
```suggestion
            This argument is deprecated. Use ``--optim adafactor`` instead. Whether or not to use the
```

sgugger commented on src/transformers/training_args.py:
```diff
@@ -807,6 +813,13 @@ def __post_init__(self):
+
+        if self.adafactor:
+            warnings.warn(
+                "`adafactor` is deprecated and will be removed in version 5 of 🤗 Transformers. Use `optim` instead",
```
```suggestion
                "`adafactor` is deprecated and will be removed in version 5 of 🤗 Transformers. Use `--optim adafactor` instead",
```

sgugger reviewed (commented):
Thanks for implementing this. I left some nits and there is one line missing which shows we need a test of this new feature :-)

I'm also wondering if we shouldn't create an Enum for all optimizer values that are supported?

manuelciosici committed:
Update src/transformers/optimization.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

manuelciosici committed:
Update src/transformers/optimization.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

manuelciosici committed:
Update src/transformers/optimization.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

manuelciosici committed:
Update src/transformers/optimization.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

manuelciosici committed:
Update src/transformers/training_args.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

manuelciosici committed:
Update src/transformers/training_args.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

manuelciosici commented:
@sgugger Thank you for the suggestions. I think an enum makes sense. I will add that and fix the missing `optimizer_cls` line.

stas00 commented on src/transformers/trainer.py:
```diff
@@ -818,17 +818,43 @@ def create_optimizer(self):
             ]
-            optimizer_cls = Adafactor if self.args.adafactor else AdamW
-            if self.args.adafactor:
+            if self.args.adafactor and self.args.optim not in {"adamw_hf", "adafactor"}:
```
this logic is unnecessary complicated, IMHO

1. deprecate --adafactor in favor of --optim adafactor
2. if `--adafactor` is passed set --optim adafactor

now you no longer need to ever consider `self.args.adafactor`

stas00 commented on src/transformers/trainer.py:
```diff
@@ -818,17 +818,43 @@ def create_optimizer(self):
+
+            # TODO the following code is a good candidate for PEP 622 once Python 3.10 becomes the
+            #  minimum required version. See, https://www.python.org/dev/peps/pep-0622/
+            if (self.args.adafactor and self.args.optim == "adamw_hf") or self.args.optim == "adafactor":
```
how can this be possible? I think my proposal above will remove the need for this.

stas00 commented on src/transformers/training_args.py:
```diff
@@ -807,6 +813,13 @@ def __post_init__(self):
+                "`adafactor` is deprecated and will be removed in version 5 of 🤗 Transformers. Use `--optim adafactor` instead",
+                FutureWarning,
+            )
+
```
right, so here, before the deprecation warning, you set `self.optim` to `adafactor` and simplify the logic in the trainer above in the PR

sgugger replied:
Good point!

manuelciosici replied:
Yes! This is much cleaner.

stas00 commented on src/transformers/training_args.py:
```diff
@@ -646,6 +648,10 @@ class TrainingArguments:
     )
+    optim: str = field(
+        default="adamw_hf",
+        metadata={"help": "The optimizer to use: adamw_hf, adamw_torch, adafactor, or apex_fused_adam."},
```
```suggestion
        metadata={"help": "The optimizer to use: adamw_hf, adamw_torch, apex_fused_adam or adafactor"},
```

suggesting to regroup `adamw`s together.

additionally, why are we not using the same naming convention for apex?

`adamw_fused_apex` perhaps?

stas00 commented on src/transformers/training_args.py:
```diff
@@ -807,6 +813,13 @@ def __post_init__(self):
             if not (self.sharded_ddp == "" or not self.sharded_ddp):
                 raise ValueError("sharded_ddp is not supported with bf16")
+
+        if self.adafactor:
+            warnings.warn(
+                "`adafactor` is deprecated and will be removed in version 5 of 🤗 Transformers. Use `--optim adafactor` instead",
+                FutureWarning,
+            )
```
this is super-confusing to a user, there are 2 unrelated deprecations here.

1. the use of adafactor optimizer is deprecated
2. the use of `--adafactor` flag is deprecated over `--optim adafactor`

So these should be 2 different depredations. Does it make sense?

if my proposal to fix `self.optim = "adafactor"` is accepted then it should be possible.

stas00 reviewed (changes requested):
Thank you for working on it, @manuelciosici 

I made a few proposals about consistency 

Please let me know if it's not clear.

(and I edited/removed my earlier comments on me thinking that the actual adafactor optimizer being deprecated, which was my reading the code incorrectly).

stas00 commented:
Additionally, let's rethink the removal of `adafactor` - this is not a great idea, IMHO, since this is a super popular and important optimizer. So how will users that rely on HF Trainer will now do the training?

While @sgugger suggested that we don't want to maintain optimizers, as it's not `transformers`, since we are adding support for a bunch of external optimizers we should do the same for `adafactor`. Let's find a good implementation (our was originally a copy from fairseq I believe) and do the same as we did with apex Adam.

The only trouble is that it won't have all the fixes we applied to make `get_lr` working.

Let me know if you'd rather discuss this in a different issue.

The intention of this PR was to deal with AdamW deprecation and not delve into adafactor, so I'm not sure why we are mixing it in.

stas00 commented:
@manuelciosici. have we lost you or are you just on an extended vacation? 

Let's finish this up, as I also was told that `apex.optimizers.FusedAdam` is an even faster fused optimizer than torch's going to benchmark it.

stas00 committed:
Update src/transformers/training_args.py

stas00 committed:
Merge remote-tracking branch 'origin/master' into deprecate_adamw

stas00 committed:
fix style

stas00 committed:
fix

stas00 commented:
I'm working on a neat HF Trainer benchmarking tool, so here it is applied to the changes introduced by this PR:

| Variation               |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:------------------------|------------------------------------:|------------:|----------------:|
| --optim adamw_hf        |                             117.544 |         132 |         2.19851 |
| --optim adamw_torch     |                             112.951 |         127 |         2.19829 |
| --optim adafactor       |                              89.194 |         100 |         2.20484 |
| --optim apex_fused_adam |                             126.232 |         142 |         2.19832 |

So torch's AdamW appears to be even slower than ours. So clearly apex's AdamW is the way to go speed-wise.

Reproducibility and other info:

```
Datetime    : 2021-12-28 20:56:42

Software:
transformers: 4.16.0.dev0
torch       : 1.10.1
cuda        : 11.3
python      : 3.8.11

Hardware:
1 GPUs      : NVIDIA GeForce RTX 3090, 23.70GB

The benchmark command line was:

CUDA_VISIBLE_DEVICES=0 python \
/hf/transformers-trainer-benchmark/scripts/benchmark/trainer-benchmark.py \
--base-cmd \
' \
examples/pytorch/translation/run_translation.py --model_name_or_path t5-base --output_dir output_dir \
--do_train --label_smoothing 0.1 --logging_strategy no --save_strategy no --per_device_train_batch_size 16 \
--max_source_length 512 --max_target_length 512 --num_train_epochs 1 --overwrite_output_dir \
--source_lang en --target_lang ro --dataset_name wmt16 --dataset_config "ro-en" \
--source_prefix "translate English to Romanian: "  --warmup_steps 50 \
--max_train_samples 10000 --dataloader_num_workers 2 \
' \
--target-metric-key train_samples_per_second --repeat-times 1 --variations \
'--optim adamw_hf|--optim adamw_torch|--optim adafactor|--optim apex_fused_adam' \
--table-format github --report-metric-keys train_loss

```

manuelciosici committed:
Change --adafactor to --optim adafactor

manuelciosici committed:
Regroup adamws together

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

manuelciosici committed:
Use Enum for optimizer values

manuelciosici commented:
@stas00 You have not lost me. I've been on vacation and before that, I accidentally messed up my GitHub notification settings, so I didn't know you had also reviewed the code. I applied all the suggestions from you.

I also changed to an `Enum` for optimizer string values as @sgugger suggested.

Let me know if I should change anything else.

manuelciosici committed:
fixup! Change --adafactor to --optim adafactor

manuelciosici committed:
fixup! Change --adafactor to --optim adafactor

manuelciosici committed:
fixup! Change --adafactor to --optim adafactor

manuelciosici committed:
Merge branch 'master' into deprecate_adamw

manuelciosici commented:
@stas00 Let me know if I should squash all the commits to a single commit so they don't pollute `master`'s commit log.

manuelciosici committed:
fixup! Use Enum for optimizer values

stas00 commented:
> @stas00 You have not lost me. [...]

Great! I will review your edits today and we can finalize it then.

> @stas00 Let me know if I should squash all the commits to a single commit so they don't pollute `master`'s commit log.

No. We always squash on merge

stas00 commented on src/transformers/training_args.py:
```diff
@@ -327,8 +338,10 @@ class TrainingArguments:
+            The optimizer to use: adamw_hf, adamw_torch, apex_fused_adam, or adafactor.
         adafactor (`bool`, *optional*, defaults to `False`):
-            Whether or not to use the [`Adafactor`] optimizer instead of [`AdamW`].
+            This argument is deprecated. Use `optim` instead.
```
```suggestion
            This argument is deprecated. Use `--optim adafactor` instead.
```

stas00 commented on src/transformers/training_args.py:
```diff
@@ -641,6 +654,10 @@ class TrainingArguments:
     )
+    optim: str = field(
+        default=OptimizerOptions.ADAMW_HF.value,
+        metadata={"help": f'The optimizer to use: {", ".join([e.value for e in OptimizerOptions])}'},
```
There is `choices` field arg for that, see other entries for examples.

stas00 commented on src/transformers/training_args.py:
```diff
@@ -809,6 +826,14 @@ def __post_init__(self):
                 raise ValueError("sharded_ddp is not supported with bf16")
+
+        if self.adafactor:
+            warnings.warn(
```
we use `logger.warning` for most warnings so let's swap it.

manuelciosici replied:
1. Are you sure about `logger.warning`? I followed the other examples in `training_args.py`:

https://github.com/manuelciosici/transformers/blob/7e62da916f392670e8e7151ed25b5cbdcb8b5a92/src/transformers/training_args.py#L759-L763

https://github.com/manuelciosici/transformers/blob/7e62da916f392670e8e7151ed25b5cbdcb8b5a92/src/transformers/training_args.py#L810-L814

https://github.com/manuelciosici/transformers/blob/7e62da916f392670e8e7151ed25b5cbdcb8b5a92/src/transformers/training_args.py#L895-L899

There are a few others.

2. If yes to 1, should I also change the other calls to `warnings.warn` to `logger.warning`?

stas00 replied:
No, you're right. There are only 2 logger.warnings for deprecated API, so it's those then should be fixed for consistency, but in a different PR.

Thank you for analyzing that file for patterns, @manuelciosici

stas00 commented on src/transformers/optimization.py:
```diff
@@ -297,7 +300,14 @@ def __init__(
+        if not no_deprecation_warning:
+            warnings.warn(
+                "This implementation of AdamW is deprecated and will be removed in a future version. Use the"
+                "PyTorch implementation torch.optim.AdamW instead.",
```
```suggestion
                "PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning",
```

stas00 commented on src/transformers/training_args.py:
```diff
@@ -69,6 +69,17 @@ def default_logdir() -> str:
     return os.path.join("runs", current_time + "_" + socket.gethostname())
 
 
+class OptimizerOptions(Enum):
```
`OptimizerNames` perhaps? 

These are definitely not optimizer options.

stas00 reviewed (changes requested):
I left a few small suggestions 

So the only remaining change is to normalize the names of the optimizers. I suggest they all start with `adam_`, including apex, so e.g. `adam_fused_apex` or `adam_apex_fused` (prefer the latter, as it goes well with `adam_torch`/`adam_hf` - i.e. having "framework/package name" in the same position.) 

And the options should be renamed as well.

We should only decide whether to drop W, as these days adam is almost always adamW.

So one of two:

1.  `adam_*`
2. `adamw_*`

The first one is probably simpler. @sgugger, what do you prefer?

manuelciosici committed:
Improved documentation for --adafactor

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

manuelciosici committed:
Add mention of no_deprecation_warning

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

manuelciosici committed:
Rename OptimizerOptions to OptimizerNames

manuelciosici committed:
Use choices for --optim

manuelciosici committed:
Move optimizer selection code to a function and add a unit test

manuelciosici committed:
Change optimizer names

manuelciosici commented:
I implemented @stas00 's suggestions (except for the `warnings.warn` bit, see the discussion above). I also remembered that @sgugger suggested adding a unit test for the optimizer selection, so I added that too.

stas00 commented on src/transformers/trainer.py:
```diff
@@ -819,17 +819,9 @@ def create_optimizer(self):
-                }
-            optimizer_kwargs["lr"] = self.args.learning_rate
+
+            optimizer_cls, optimizer_kwargs = Trainer.get_optimizercls_and_params(self.args)
```
```suggestion
            optimizer_cls, optimizer_kwargs = Trainer.get_optimizer_cls_and_kwargs(self.args)
```

stas00 commented on src/transformers/trainer.py:
```diff
@@ -844,6 +836,48 @@ def create_optimizer(self):
         return self.optimizer
 
+    @staticmethod
+    def get_optimizercls_and_params(args: TrainingArguments) -> Tuple[Any, Any]:
```
```suggestion
    def get_optimizer_cls_and_kwargs(args: TrainingArguments) -> Tuple[Any, Any]:
```

stas00 commented on src/transformers/trainer.py:
```diff
@@ -844,6 +836,48 @@ def create_optimizer(self):
+            "betas": (args.adam_beta1, args.adam_beta2),
+            "eps": args.adam_epsilon,
+        }
+        # TODO the following code is a good candidate for PEP 622 once Python 3.10 becomes the
+        #  minimum required version. See, https://www.python.org/dev/peps/pep-0622/
```
this is very very unlikely to happen any time soon (like years), so just as well let's not add it?
```suggestion
```

stas00 commented on tests/test_trainer.py:
```diff
@@ -1690,3 +1692,85 @@ def hp_name(trial):
+        """
+        Checks that the common case for an optimizer works.
+        """
+        actual_cls, optim_params = Trainer.get_optimizercls_and_params(args)
```
```suggestion
        actual_cls, optim_kwargs = Trainer.get_optimizer_cls_and_kwargs(args)
```

stas00 commented on tests/test_trainer.py:
```diff
@@ -1690,3 +1692,85 @@ def hp_name(trial):
+        """
+        actual_cls, optim_params = Trainer.get_optimizercls_and_params(args)
+        self.assertEqual(expected_cls, actual_cls)
+        self.assertIsNotNone(optim_params)
```
```suggestion
        self.assertIsNotNone(optim_kwargs)
```

stas00 commented on tests/test_trainer.py:
```diff
@@ -1690,3 +1692,85 @@ def hp_name(trial):
+
+        self.check_optim(args, mandatory_params, AdamW)
+
+    def test_adam_torch(self):
```
This is almost identical to the previous test - use `parameterized` and have 2 tests with a single body?

This could include the real apex test as well, except it'd need to skip if `is_apex_available` is `False`.

```
from transformers.file_utils import is_apex_available
...
```

stas00 commented on tests/test_trainer.py:
```diff
@@ -1690,3 +1692,85 @@ def hp_name(trial):
+
+        self.check_optim(args, mandatory_params, AdamW)
+
+    def test_fused_adam(self):
```
awesome tests, @manuelciosici - but where is the actual `apex` test when it's installed? :) I suggest to add it earlier - see my comment above.

stas00 reviewed (changes requested):
Very excellent work, @manuelciosici 

One nit on the new helper method - really just needed `_` to make it readable. Wasn't sure if param or kwargs would be more intuitive to the user - I just followed your assignments naming to be consistent.

Great tests - suggesting to add a missing apex-exists test and refactor 3 almost identical tests into 1 parametrized test.

manuelciosici committed:
Rename method

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

manuelciosici committed:
Rename method

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

manuelciosici committed:
Remove TODO comment

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

manuelciosici committed:
Rename variable

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

manuelciosici committed:
Rename variable

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

manuelciosici committed:
Rename function

manuelciosici committed:
Rename variable

manuelciosici committed:
Parameterize the tests for supported optimizers

manuelciosici committed:
Refactor

manuelciosici commented:
@stas00 I managed to parameterize the tests for AdamW (HF and torch) and Adafactor and refactored the testing code.

For testing `apex.optimizers.FusedAdam`, since `Trainer.get_optimizer_cls_and_kwargs` does not use `FusedAdam` but only returns a class called `apex.optimizers.FusedAdam`, I thought I could mock `FusedAdam`'s  presence or absence independent of an `apex` installation.

In `test_fused_adam`, I mock `apex.optimizers.FusedAdam` and check that `Trainer.get_optimizer_cls_and_kwargs` returns the object mocking `apex.optimizers.FusedAdam`.

In `test_fused_adam_no_apex`, I simulate a missing `apex.optimizers.FusedAdam` by setting the `apex` namespace to `None` so `Trainer.get_optimizer_cls_and_kwargs` cannot import `apex.optimizers.FusedAdam` even if `apex` is installed.

Does this testing approach make sense? Or should I write `apex`-dependent tests and use `is_apex_available`?

manuelciosici committed:
Attempt to make tests pass on CircleCI

stas00 commented:
> @stas00 I managed to parameterize the tests for AdamW (HF and torch) and Adafactor and refactored the testing code.
> 
> For testing `apex.optimizers.FusedAdam`, since `Trainer.get_optimizer_cls_and_kwargs` does not use `FusedAdam` but only returns a class called `apex.optimizers.FusedAdam`, I thought I could mock `FusedAdam`'s presence or absence independent of an `apex` installation.

Perhaps we are looking at different code?  `Trainer.get_optimizer_cls_and_kwargs` branches are almost identical for torch or apex. What do you mean it only returns a class. It returns kwargs too.

> In `test_fused_adam`, I mock `apex.optimizers.FusedAdam` and check that `Trainer.get_optimizer_cls_and_kwargs` returns the object mocking `apex.optimizers.FusedAdam`.
> 
> In `test_fused_adam_no_apex`, I simulate a missing `apex.optimizers.FusedAdam` by setting the `apex` namespace to `None` so `Trainer.get_optimizer_cls_and_kwargs` cannot import `apex.optimizers.FusedAdam` even if `apex` is installed.
> 
> Does this testing approach make sense? Or should I write `apex`-dependent tests and use `is_apex_available`?

Your testing is fabulous, Manuel. But you're not testing the rest thing when it's available. What if apex does a breaking change and we are still mocking what it should be doing and not what it does. Hence, I'm suggesting that there should be an identical test to torch's except need to inject a skip in that parameterized test if  `testing apex and not is_apex_available()`. Would it'd be easier if I wrote it?

manuelciosici committed:
Add a test with apex

manuelciosici commented:
@stas00 

>  What do you mean it only returns a class. It returns kwargs too.

You are right, it returns a class and the keyword arguments. I didn't express myself properly.

I added a new test that uses `FusedAdam` from `apex` if `is_apex_available` is `True`. I couldn't figure out how to test using a parameterization of `test_supported_optim` since I couldn't figure out how to add the `FusedAdam` class to `test_params` without causing the code to throw a `ModuleNotFoundError` in environments without `apex`. In other words, I can't figure out how to conditionally add to `test_params` so that `@parameterized.expand` still works. If there is a way, can you show me?

stas00 committed:
rework to add apex to parameterized; add actual train test

stas00 commented:
Done.

Also added an actual `train()` in the test, which is the ultimate test. 

Note that I get the last test fail:

```
tests/test_trainer.py .F

___________________________________________________ TrainerOptimizerChoiceTest.test_fused_adam_no_apex ___________________________________________________

self = <tests.test_trainer.TrainerOptimizerChoiceTest testMethod=test_fused_adam_no_apex>

    def test_fused_adam_no_apex(self):
        args = TrainingArguments(optim=OptimizerNames.ADAM_APEX_FUSED.value, output_dir="None")
    
        # Pretend that apex does not exist, even if installed. By setting apex to None, importing
        # apex will fail even if apex is installed.
        with patch.dict("sys.modules", {"apex": None}):
            with self.assertRaises(ValueError):
>               Trainer.get_optimizer_cls_and_kwargs(args)
E               AssertionError: ValueError not raised

tests/test_trainer.py:1795: AssertionError

tests/test_trainer.py .....   
```

I'm not an expert on Mock, so not sure why it doesn't override like it should.

I have apex installed when I run this test.

Can you reproduce it?

stas00 committed:
fix import when torch is not available

stas00 committed:
fix optim_test_params when torch is not available

stas00 committed:
fix optim_test_params when torch is not available

stas00 committed:
re-org

stas00 committed:
small re-org

stas00 committed:
fix test_fused_adam_no_apex

stas00 reviewed (approved):
This looks good now. Thank you for bearing with me, Manuel.

So just need to get @sgugger's approval and we are good to merge this. (he's back on Jan 10th, so we will have to wait a week).

manuelciosici commented:
@stas00 Thank you for the code cleanup, figuring out the mocking issue, and for your patience. This PR has been an educational experience (I didn't know about `parameterized`). I'm looking forward to figuring out what my next contribution should be. Let me know if you have any suggestions.

stas00 commented:
Since clearly you're interesting in easier testing, you may find some useful tidbits in this extensive doc: https://huggingface.co/docs/transformers/testing, e.g. for different parameterization situations https://huggingface.co/docs/transformers/testing#parametrization

optimizers-wise I think the next interesting but challenging thing to add is BNB 8bit optimizer
https://github.com/huggingface/transformers/issues/14819 but we are still discussing how it'd work.

The other thing to potentially experiment with is 
https://www.deepspeed.ai/tutorials/onebit-adam/
but I haven't had a chance to understand it so I have no idea whether it can be used outside of Deepspeed or just with deepspeed.

sgugger commented on src/transformers/trainer.py:
```diff
@@ -844,6 +836,46 @@ def create_optimizer(self):
+            "betas": (args.adam_beta1, args.adam_beta2),
+            "eps": args.adam_epsilon,
+        }
+        if args.optim == OptimizerNames.ADAFACTOR.value:
+            optimizer_cls = Adafactor
+            optimizer_kwargs.update({"scale_parameter": False, "relative_step": False})
+        elif args.optim == OptimizerNames.ADAM_HF.value:
+            from .optimization import AdamW
+
+            optimizer_cls = AdamW
+            optimizer_kwargs.update(adam_kwargs)
+        elif args.optim == OptimizerNames.ADAM_TORCH.value:
+            from torch.optim import AdamW
+
+            optimizer_cls = AdamW
+            optimizer_kwargs.update(adam_kwargs)
+        elif args.optim == OptimizerNames.ADAM_APEX_FUSED.value:
```
This is a very use of an enum. The `args.optim` attributed should be converted to the enum type (you just need to set it with a type `OptimizerNames` instead of `str` and it will accept both `str` and `OptimizerNames` values as well as doing the conversion in the post init) and those tests should have all the `.value` removed.

manuelciosici replied:
Thank you. I've updated everything.

sgugger commented on src/transformers/training_args.py:
```diff
@@ -641,6 +654,10 @@ class TrainingArguments:
     label_smoothing_factor: float = field(
         default=0.0, metadata={"help": "The label smoothing epsilon to apply (zero means no label smoothing)."}
     )
+    optim: str = field(
```
```suggestion
    optim: OptimizerNames = field(
```

sgugger commented on src/transformers/training_args.py:
```diff
@@ -809,6 +826,14 @@ def __post_init__(self):
+                "`--adafactor` is deprecated and will be removed in version 5 of 🤗 Transformers. Use `--optim adafactor` instead",
+                FutureWarning,
+            )
+            self.optim = OptimizerNames.ADAFACTOR.value
```
```suggestion
            self.optim = OptimizerNames.ADAFACTOR
```

sgugger reviewed (approved):
Thanks for waiting for me! This looks great (nice new tests!) but I have a few last comments.

manuelciosici committed:
Update src/transformers/training_args.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

manuelciosici committed:
Update src/transformers/training_args.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

manuelciosici committed:
Update src/transformers/training_args.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

stas00 commented:
I think the only other remaining item that I didn't hear you weigh on, @sgugger, is whether we should call these `--adam_foo` or `--adamw_foo` since the class names are `AdamW` (except apex).

manuelciosici committed:
Remove .value from OptimizerNames

sgugger commented:
Oh sorry, I didn't catch that. It should adamw everywhere IMO.

stas00 commented:
> Oh sorry, I didn't catch that. It should adamw everywhere IMO.

Thank you for validating that, @sgugger. 

@manuelciosici, so one more tiny change please `s|--adam_|--adamw_|` 

Thank you!

manuelciosici committed:
Rename optimizer strings s|--adam_|--adamw_|

manuelciosici commented:
I removed `.value` from everywhere and ensured that the tests pass. I have also changed optimizer name strings as @stas00  asked.

Let me know if I should change anything else.

stas00 commented:
but the rest needs to updated to match, e.g. currently many torch tests fail with:
```
E           ValueError: Trainer cannot instantiate unsupported optimizer: adamw_hf
```

manuelciosici committed:
Also rename Enum options

manuelciosici commented:
@stas00 I just saw that. I'm trying to figure out what I misunderstood.

stas00 committed:
small fix

manuelciosici committed:
Fix instantiation of OptimizerNames. Remove redundant test

manuelciosici committed:
Use ExplicitEnum instead of Enum

manuelciosici commented:
@stas00 I surprised that fixes it. On my end, I just fixed it by adding `self.optim = OptimizerNames(self.optim)` in post-init. I also had to remove a now redundant unit test.

stas00 commented:
It's the same as:
```
elif args.optim == OptimizerNames.ADAMW_HF:
```

manuelciosici committed:
Add unit test with string optimizer

manuelciosici commented:
@stas00 Commit e73249ce130bc81c392f5b9a3224f5ba64a2214d makes the unit tests pass, but it doesn't work when `--optim` is explicitly set on the command line. `TrainingArguments` does not automatically convert strings to enums. For a parallel, see https://github.com/huggingface/transformers/blob/9a94bb8e218033cffa1ef380010b528410ba3ca7/src/transformers/training_args.py#L755

So, calling a script on the command line (for example, `examples/pytorch/language-modeling/run_clm.py`) with any explicit `--optim` (even if only with `--optim adamw_hf`), throws an error from `Trainer.get_optimizer_cls_and_kwargs` since `get_optimizer_cls_and_kwargs` receives a string that does not match any of the if branches which test for `Enum` values.

I added another test for specifying the optimizer name as a string. Also, I removed `test_optim_unsupported` since, with these commits, `--optim` no longer accepts strings that are not in `OptimizerNames`.

I also changed the default value from `OptimizerNames.ADAMW_HF` to `OptimizerNames.ADAMW_HF.value`. With `default=OptimizerNames.ADAMW_HF` calling `--help` on the CLI gives:

```
  --optim {adamw_hf,adamw_torch,adamw_apex_fused,adafactor}
                        The optimizer to use. (default:
                        OptimizerNames.ADAMW_HF)
```

While `default=OptimizerNames.ADAMW_HF.value` gives

```
  --optim {adamw_hf,adamw_torch,adamw_apex_fused,adafactor}
                        The optimizer to use. (default: adamw_hf)
```

The first one leaks the internal object name, while the second indicates the string we want users to pass.

Finally, I make `OptimizerNames` inherit `ExplicitEnum` instead of `Enum` because I saw `SchedulerType` do the same and it seems more elegant.
https://github.com/huggingface/transformers/blob/9a94bb8e218033cffa1ef380010b528410ba3ca7/src/transformers/trainer_utils.py#L281

stas00 commented on src/transformers/training_args.py:
```diff
@@ -641,6 +655,10 @@ class TrainingArguments:
         default=0.0, metadata={"help": "The label smoothing epsilon to apply (zero means no label smoothing)."}
     )
+    optim: OptimizerNames = field(
+        default=OptimizerNames.ADAMW_HF.value,
```
@sgugger, what would you like here? Everywhere else with enum we set the default to the actual string, so here it can be:
- `OptimizerNames.ADAMW_HF.value`
- `adamw_hf`

Use the latter for consistency?

sgugger replied:
I personally prefer the string values, yes.

manuelciosici replied:
@stas00 @sgugger I changed the default to the string value.

manuelciosici committed:
Change optimizer default to string value

sgugger reviewed (approved):
All looking great on my side! @stas00 I'll let you merge if you're happy as well.

Thanks again for all your work on this @manuelciosici !

stas00 commented:
I second that - thank you, @manuelciosici!

manuelciosici commented:
@stas00 @sgugger Thank you for guiding me!
